guides/inventing-service.md

```diff
@@ -0,0 +1,856 @@
+
+        Box::pin(async move {
+            let mut response = this.inner_handler.call(request).await?;
+            response.set_header("Content-Type", "application/json");
+            Ok(response)
+        })
+    }
+}
+```
+
+Finally the `Handler` passed to `Server::run` must use `Request` and `Response`:
+
+```rust
+impl Server {
+    async fn run<T>(self, mut handler: T) -> Result<(), Error>
+    where
+        T: Handler<Request, Response = Response>,
+    {
+        // ...
+    }
+}
+```
+
+Creating the server has not changed:
+
+```rust
+let handler = RequestHandler;
+let handler = Timeout::new(handler, Duration::from_secs(30));
+let handler = JsonContentType::new(handler);
+
+server.run(handler).await
+```
+
+So, we now have a `Handler` trait that makes it possible to break our
+application up into small independent parts and re-use them. Not bad!
+
+## "What if I told you..."
+
+Until now we've only talked about HTTP from the servers perspective. But our
+`Handler` trait actually fits HTTP clients as well. One can imagine a client
+`Handler` that accepts some `Request` struct and asynchronously sends it to
+someone on the internet. Our `Timeout` wrapper is actually useful here as well.
+`JsonContentType` probably isn't since its not the clients job to set response
+headers.
+
+Since our `Handler` trait is useful for defining both servers and clients,
+`Handler` probably isn't an appropriate name anymore. A client doesn't handle an
+request, it sends it to a server and the server handles it. Lets instead call
+our trait `Service`:
+
+```rust
+trait Service<Request> {
+    type Response;
+    type Error;
+    type Future: Future<Output = Result<Self::Response, Self::Error>>;
+
+    fn call(&mut self, request: Request) -> Self::Future;
+}
+```
+
+This is actually _almost_ the `Service` trait defined in Tower. If you've been
+able to follow along until this point you now understand most of Tower. Besides
+the `Service` trait Tower also provides several utilities that implement
+`Service` by wrapping some other type that also implements `Service`. Exactly
+like we did with `Timeout` and `JsonContentType`. Things can be composed in ways
+similar to what we've just did.
+
+Some example services provided by Tower:
+
+- `Timeout` - This is pretty much identical to the timeout we have built.
+- `Retry` - To automatically retry failed requests.
+- `RateLimit` - To limit the number of requests a service can handle over a
+  period of time.
+
+Types like `Timeout` and `JsonContentType` are typically called "middlewares"
+since they wrap another `Service` and transform the request or response in some
+way. Whereas types like `RequestHandler` is typically called "leaf services"
+since it sits at the leaves of a tree of nested services. The actual responses
+are normally produced in leaf services and modified by middlewares.
+
+The only thing left to talk about is "backpressure" and `poll_ready`.
+
+## Backpressure
```

Comment:
I'd recommend adding a `backpressure` explanation article here. Such as this wonder one: https://medium.com/@jayphelps/backpressure-explained-the-flow-of-data-through-software-2350b3e77ce7.

This would help a lot to those new to `backpressure`. What do you think?

Reply:
Good idea! I added a link in 7461ae492ddaa